### PR TITLE
Email verification code

### PR DESCRIPTION
**Pull request by HexelDev**
Email verification code.

Issue #1311

### CODE REVIEW

File: src/main/java/fr/xephi/authme/command/executable/verification/VerificationCommand.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    @Inject
+    private VerificationCodeManager codeManager;
+
+    @Override
+    public void runCommand(Player player, List<String> arguments) {
+        final String playerName = player.getName();
+
+        if (!codeManager.isEnabled()) {
+            ConsoleLogger.warning("Mail API is not set");
+            commonService.send(player, MessageKey.INCOMPLETE_EMAIL_SETTINGS);
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/command/executable/verification/VerificationCommand.java]: This is not necessarily true. I'll take care of this (dropping a comment so I don't forget)

**Comment by ljacqu** [comment]
@HexelDev, thanks for doing this! You did a great job at implementing this. This was a big feature to add with a lot of unclear details. You faced these changing assumptions / requirements head on and produced a solid implementation.

What I changed before merging was mainly moving the check for the permission within the verification code service so I can just call one method when I just care about answering "is a verification code required or no?". Besides that I removed the commented out image code and would vote not to implement that unless someone requests it.

Outlook:
- We should discuss about error handling. If an email cannot be sent (`EmailService#sendVerificationMail` returns a boolean indicating success) we do not handle that in the verification code service. This blocks a user from being able to do anything until emails can be sent.
- I think we could revisit the methods on `RecoveryCodeService`. From the naming, `isVerificationRequired` and `isCodeRequired` kind of sounds like the same to me. I also wonder if we shouldn't move more of the logic into that service instead of calling a handful of methods in `VerificationCommand` (I would expect to just call `verificationService#checkCode` or something).
- Could we add `/verify` also as label? Would be in line with our `/register`, `/login`, etc. commands

**Comment by HexelDev** [comment]
> We should discuss about error handling. If an email cannot be sent (EmailService#sendVerificationMail returns a boolean indicating success) we do not handle that in the verification code service. This blocks a user from being able to do anything until emails can be sent.

We could add an error message that notify the user about this issue

> I think we could revisit the methods on RecoveryCodeService. From the naming, isVerificationRequired and isCodeRequired kind of sounds like the same to me. 

We could handle it better, that was my main "issue" during the coding process

- `isVerificationRequired` should define if the user is able to access the verification process
  - checks if the _emails are enabled_, the _user has a real email_, the _user has the permission_ and the _user has not been verified yet_.
- `isCodeRequired` should define if the user has a code and must use it to verify his identity
  - checks if the _user has a code_ and has not been verified yet.

> I also wonder if we shouldn't move more of the logic into that service instead of calling a handful of methods in VerificationCommand (I would expect to just call verificationService#checkCode or something).

I'm not a lot into the project, but as I saw in other commands (in a more in a more synthetic way) the messages where handled there, providing a good visual feedback, in my opinion, but obviously it's more a feedback than a smooth code...

> Could we add /verify also as label? Would be in line with our /register, /login, etc. commands

Yeah, that was the idea... I forgot to implement it ^.^

**Comment by HexelDev** [comment]
@ljacqu Should we open an issue for this improvements?

**Comment by ljacqu** [comment]
@HexelDev If you could, that would be great. I'm especially interest in the error handling.

**Comment by HexelDev** [comment]
@ljacqu Ok, I'll create a new issue this evening

**Pull request merged**